[droidcamsrc] implement camerabin2's preview-image interface

peat-psuwit opened the pull request:
This patch implement camerabin2's preview-image interface,
which make the element post a preview-image bus message when
post-previews property is enabled.

As Android doesn't seem to have the concept of preview-image (there are
raw & postview callbacks but they seem not to be called), I replicate
what Qt does with it's native Android plugin, which is storing the
latest viewfinder frame and using that as the preview. This require
changing how the preview callback flags is calculated as otherwise
preview callback won't be called. The last preview buffer is kept
within the structure of dev but use the seperated lock and condition
variable, as the buffer is updated in the callback called from Android
side.

Camerabin2's preview-image interface also require that the element has
the preview-caps property and must send the preview-image according to
this caps. To fulfill that, a camerabinpipeline is constructed inside
the gstdroidcamsrc element. This pipeline will take care of all
conversion required to make the preview image comply to the provided
caps and will also actually posting it to the bus.

When the image capture is requested, the post-preview property is
copied to image capturing structure so that the change after the
capture won't affect this capture. When the shutter callback is
called and if the preview image is requested, the kept viewfinder
buffer is sent to the preview pipeline for processing & posting. For
video recording, preview is posted when the recording is requested.

This can remove the requirement to patch QtMultimedia not to set the
post-preview flag, and improve the correctness of the element outside
of Mer and Sailfish OS.

abranson commented:
There are additional camera messages on the droidmedia side which aren't hooked up to send the data to callbacks. I'm not sure why they're present at all in this state, but I tried one out recently while trying to get HDR working:

https://github.com/sailfishos/droidmedia/commit/5f7e3aee5f7cc71d3213f151da88b87edaac6bb4

 If this would be a simpler route to what you're trying to do, we could merge it or add more.

peat-psuwit commented:
`CAMERA_MSG_RAW_IMAGE_NOTIFY` is actually a data-less variant of `CAMERA_MSG_RAW_IMAGE`. In Android, it's used when raw image callback is available but no buffer is provided to the Camera class. Thus, it's useless in this use-case.

https://github.com/LineageOS/android_frameworks_base/blob/lineage-17.0/core/jni/android_hardware_Camera.cpp#L905

The actual problem is that I can't find any way to know whether the device will fire postview or raw callback or not. Meanwhile, decoding the just-encoded jpeg to be used as a preview image seems to be inefficient. Thus, I guess the preview buffers received from preview callback is the best way to implement this.

peat-psuwit committed:
[droidcamsrc] implement camerabin2's preview-image interface

This patch implement camerabin2's preview-image interface,
which make the element post a preview-image bus message when
post-previews property is enabled.

As Android doesn't seem to have the concept of preview-image (there are
raw & postview callbacks but they seem not to be called), I replicate
what Qt does with it's native Android plugin, which is storing the
latest viewfinder frame and using that as the preview. This require
changing how the preview callback flags is calculated as otherwise
preview callback won't be called. The last preview buffer is kept
within the structure of dev but use the seperated lock and condition
variable, as the buffer is updated in the callback called from Android
side.

Camerabin2's preview-image interface also require that the element has
the preview-caps property and must send the preview-image according to
this caps. To fulfill that, a camerabinpipeline is constructed inside
the gstdroidcamsrc element. This pipeline will take care of all
conversion required to make the preview image comply to the provided
caps and will also actually posting it to the bus.

When the image capture is requested, the post-preview property is
copied to image capturing structure so that the change after the
capture won't affect this capture. When the shutter callback is
called and if the preview image is requested, the kept viewfinder
buffer is sent to the preview pipeline for processing & posting. For
video recording, preview is posted when the recording is requested.

This can remove the requirement to patch QtMultimedia not to set the
post-preview flag, and improve the correctness of the element outside
of Mer and Sailfish OS.

peat-psuwit commented:
So, after testing it in Ubuntu Touch for a while, I think this patch is now ready for review. @abranson?

abranson commented on gst/droidcamsrc/gstdroidcamsrc.c:
```diff
@@ -3,6 +3,7 @@
  *
  * Copyright (C) 2014 Mohammed Sameer <msameer@foolab.org>
  * Copyright (C) 2015-2016 Jolla LTD.
+ * Copyright (C) 2010 Texas Instruments, Inc
```
Where did this come from? It's in the header too.

peat-psuwit replied:
Sorry, forget to mention this.

The code for handling the preview pipeline is pretty much copied from `gstbasecamerasrc` in gst-plugins-bad [1]. You'll see that the code is copyrighted by TI and licensed under LGPL2+. Thus, re-distribute with LGPL2.1+ should be compatible.

[1] https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/blob/master/gst-libs/gst/basecamerabinsrc/gstbasecamerasrc.c

abranson replied:
Ok, useful piece of information there, thanks! Aligning with the base impl is a useful target.

abranson commented:
For reference, here's the patch in QtMultimedia that could be dropped:
https://git.sailfishos.org/mer-core/qtmultimedia/commit/770e009cff0e5eaca544e177c5aaa2e2ce6523a9

abranson commented:
Does this also require https://github.com/sailfishos/gst-droid/pull/39? That mentions a deadlock in pulling preview frames.

peat-psuwit commented:
It's a separate issue. I can't remember exactly what happens there preview-frame-wise, but without that patch, take_picture() will hang the main thread on FP2 with flash on. However, 2 PRs conflict with each other as they modify the same structure. If a PR is merged, I'll rebase another one.

abranson commented:
I finally managed to give this a test with that Qt patch dropped. Everything seemed to work fine, with the processing counter working fine and idle state occurring at the right time, which were both mentioned in that patch.

There were a lot of complaints on the console about the 'preview-filter' property not being present in droidcamsrc. That looks fairly useful in some circumstances - it adds the ability to pass the preview image through a filter on its way to the camera, presumable if a software filter will be applied to the final image. Maybe this should be present here, even as a TODO?

abranson commented on gst/droidcamsrc/gstdroidcamsrcdev.h:
```diff
@@ -3,6 +3,8 @@
  * Copyright (C) 2014 Mohammed Sameer <msameer@foolab.org>
  * Copyright (C) 2015-2016 Jolla LTD.
+ * Copyright (C) 2020 UBports foundation.
+ * Author: Ratchanan Srirattanamet <ratchanan@ubports.com>
```
We don't really do individual author comments in source files. MSameer added his own when he first wrote this module, but there have been many contributors since and they get their mentions purely in the changelog generated from git. Feel free to add yourself to the AUTHORS file if you want to, but we should keep the source headers clear of such things.

abranson replied:
I noticed you took out the UBports credit out too - I was only talking about the Author ones. If your work is supported by UBports then it would be great to show this here.

peat-psuwit replied:
Oh, OK. Will add it back.

peat-psuwit commented:
It's not hard to implement `preview-filter` property; just need to copy a bit more code from `gstcamerabasesrc`. I'll do that in a separate commit. However, this filter affects only "preview image" according to this patch, not the "viewfinder", so I'm not sure how useful that is.

abranson commented:
I guess the point is to apply any filters that would be applied to the finished image, which you can't do in real time for the viewfinder? In the case of HDR scene mode, for example, that image could be up for quite some time as all the processing gets done.

peat-psuwit committed:
[droidcamsrc] also implement preview-filter property

This property takes an element and place it inside the preview-image
pipeline.
